gitclone/gitclone_test.go

```diff
@@ -0,0 +1,411 @@
+				`git "merge" "origin/master"`,
+				`git "log" "-1" "--format=%H"`,
+				`git "fetch" "origin" "refs/heads/test/commit-messages"`,
+				`git "merge" "76a934ae80f12bb9b504bbc86f64a1d310e5db64"`,
+				`git "checkout" "--detach"`,
+			},
+		},
+		{
+			name: "Checkout PR fork - manual merge",
+			cfg: Config{
+				RepositoryURL:   "https://github.com/bitrise-io/git-clone-test.git",
+				PRRepositoryURL: "https://github.com/bitrise-io/other-repo.git",
+				Branch:          "test/commit-messages",
+				BranchDest:      "master",
+				Commit:          "76a934ae80f12bb9b504bbc86f64a1d310e5db64",
+				ManualMerge:     true,
+			},
+			want: nil,
+			wantCmds: []string{
+				`git "fetch" "origin" "refs/heads/master"`,
+				`git "checkout" "master"`,
+				`git "merge" "origin/master"`,
+				`git "log" "-1" "--format=%H"`,
+				`git "remote" "add" "fork" "https://github.com/bitrise-io/other-repo.git"`,
+				`git "fetch" "fork" "refs/heads/test/commit-messages"`,
+				`git "merge" "fork/test/commit-messages"`,
+				`git "checkout" "--detach"`,
+			},
+		},
+		{
+			name: "Checkout PR fork (private) - auto merge (overrides manual merge flag) - merge branch",
+			cfg: Config{
+				RepositoryURL:   "https://github.com/bitrise-io/git-clone-test.git",
+				PRRepositoryURL: "git@github.com:bitrise-io/other-repo.git",
+				Branch:          "test/commit-messages",
+				BranchDest:      "master",
+				PRMergeBranch:   "pull/7/merge",
+				PRID:            7,
+				Commit:          "76a934ae80f12bb9b504bbc86f64a1d310e5db64",
+				ManualMerge:     true,
+			},
+			want: nil,
+			wantCmds: []string{
+				`git "fetch" "origin" "refs/heads/master"`,
+				`git "fetch" "origin" "refs/pull/7/head:pull/7"`,
+				`git "checkout" "master"`,
+				`git "merge" "origin/master"`,
+				`git "merge" "pull/7"`,
+				`git "checkout" "--detach"`,
+			},
+		},
+		{
+			name: "Checkout PR fork (private) - auto merge (overrides manual merge flag) - fails",
+			cfg: Config{
+				RepositoryURL:   "https://github.com/bitrise-io/git-clone-test.git",
+				PRRepositoryURL: "git@github.com:bitrise-io/other-repo.git",
+				Branch:          "test/commit-messages",
+				BranchDest:      "master",
+				Commit:          "76a934ae80f12bb9b504bbc86f64a1d310e5db64",
+				ManualMerge:     true,
+			},
+			want: newStepError(
+				"auto_merge_failed",
+				fmt.Errorf("merging PR (automatic) failed: %v", "there is no Pull Request branch and can't download diff file"),
+				"Merging pull request failed",
+			),
+			wantCmds: []string{
+				`git "fetch" "origin" "refs/heads/master"`,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockRunner := newMockRunner(tt.cmdOutputs)
+			runner = mockRunner
+			got := checkoutState(git.Git{}, tt.cfg)
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("checkoutState().err = (%#v), want %#v", got, tt.want)
+			}
+			// require.Equal(t, tt.want, got)
+			if !reflect.DeepEqual(mockRunner.Cmds(), tt.wantCmds) {
```

Comment:
I'm not sure what `reflect.DeepEqual` does but if just want to assert that two slices contains the same elements in the same order we could use the [testify/assert package](https://github.com/stretchr/testify/issues/759).

Reply:
It is doing the exact same thing, except require.Equal prints out a better diff on case of failure. Do not want to introduce the testify dependency for a single test yet.